packetbeat/protos/dns/dns.go

```diff
@@ -707,3 +714,160 @@ func decodeDnsPacket(data []byte) (dns *layers.DNS, err error) {
+	}
+
+	// what kind of checks should be done here ?
+	// gopacket decodeDns already check QDCount, ANCount, NScount, ARcount
+	data := dns.messageParser(stream)
+
+	if data == nil {
+		logp.Debug("dns", NonDnsCompleteMsg+" addresses %s, length %d",
+			tcpTuple.String(), len(stream.data))
+
+		// drop this tcp stream. Will retry parsing with the next
+		// segment in it
+		priv.Data[dir] = nil
+		return priv
+	}
+
+	dns.messageComplete(tcpTuple, dir, stream, data)
+
+	return priv
+}
+
+// return decoded data so we don't have to do decode twice
+// return nil if failed
+func (dns *Dns) messageParser(s *DnsStream) *layers.DNS {
+	dnsData, err := decodeDnsData(s.data)
+
+	if err != nil {
+		return nil
+	}
+
+	// no other check ? are gopacket checks enough?
+	return dnsData
+}
+
+func (dns *Dns) messageComplete(tcpTuple *common.TcpTuple, dir uint8, s *DnsStream, decodedData *layers.DNS) {
+	dns.handleDns(s.message, tcpTuple, dir, s.data, decodedData)
+
+	s.PrepareForNewMessage()
+}
+
+func (dns *Dns) handleDns(m *DnsMessage, tcpTuple *common.TcpTuple, dir uint8, data []byte, decodedData *layers.DNS) {
+	dnsTuple := DnsTupleFromIpPort(&m.Tuple, TransportTcp, decodedData.ID)
+	m.CmdlineTuple = procs.ProcWatcher.FindProcessesTuple(tcpTuple.IpPort())
+	m.Data = decodedData
+	m.Length = len(data)
+
+	if decodedData.QR == Query {
+		dns.receivedDnsRequest(&dnsTuple, m)
+	} else /* Response */ {
+		dns.receivedDnsResponse(&dnsTuple, m)
+	}
+}
+
+func (stream *DnsStream) PrepareForNewMessage() {
+	stream.data = stream.data[stream.parseOffset:] // useful var ?
+	stream.parseOffset = 0                         // useful var ?...
+	stream.message = nil
+}
+
+func (dns *Dns) ReceivedFin(tcpTuple *common.TcpTuple, dir uint8, private protos.ProtocolData) protos.ProtocolData {
+	if private == nil {
+		return private
+	}
+	dnsData, ok := private.(dnsPrivateData)
+	if !ok {
+		return private
+	}
+	if dnsData.Data[dir] == nil {
+		return dnsData
+	}
+
+	stream := dnsData.Data[dir]
+	if stream.message != nil {
+		decodedData := dns.messageParser(stream)
+		if decodedData != nil {
+			dns.handleDns(stream.message, tcpTuple, dir, stream.data, decodedData)
+			stream.PrepareForNewMessage()
+		}
+	}
+
+	return dnsData
+}
+
+func (dns *Dns) GapInStream(tcpTuple *common.TcpTuple, dir uint8, nbytes int, private protos.ProtocolData) (priv protos.ProtocolData, drop bool) {
+	dnsData, ok := private.(dnsPrivateData)
+
+	if !ok {
+		return private, false
+	}
+
+	stream := dnsData.Data[dir]
+
+	if stream == nil || stream.message == nil {
+		return private, false
+	}
+
+	decodedData := dns.messageParser(stream)
+	if decodedData == nil {
+		// cannot create a DnsTuple because we cannot read ID from a failed decodedData
+		// -> no error message in Notes
+
+		//trans := dns.getTransaction(dnsTuple)
+		//trans.Notes = append(trans.Notes, "Packet loss while capturing")
+
+		logp.Debug("dns", "Packet loss, dropping the stream")
```

Comment:
If it was a DNS request that encountered loss then we are not going to have the data needed to send even a partial transaction.

If it was a DNS response that encountered loss then you might be able to extract the request ID from `dnsData.Data[TcpDirectionOriginal]` with some work.

Side note: At some point (not now) there should be some refactoring to remove the duplicated TCP transaction state data. It is now stored both in the cache (`dns.transactions`) and in the tcp package as `protos.ProtocolData`. If this occurred then you would not need to call `dns.getTransaction` because all the data would be in `protos.ProtocolData`.